libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxSpeechLoader.kt

```diff
@@ -17,28 +20,92 @@ import com.mapbox.navigation.ui.utils.internal.resource.ResourceLoader
 import com.mapbox.navigation.ui.utils.internal.resource.load
 import com.mapbox.navigation.ui.voice.model.TypeAndAnnouncement
 import com.mapbox.navigation.ui.voice.options.MapboxSpeechApiOptions
+import com.mapbox.navigation.utils.internal.InternalJobControlFactory
+import com.mapbox.navigation.utils.internal.logE
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
 import java.net.MalformedURLException
 import java.net.URL
 
-internal class MapboxSpeechProvider(
+internal class MapboxSpeechLoader(
     private val accessToken: String,
     private val language: String,
     private val urlSkuTokenProvider: UrlSkuTokenProvider,
     private val options: MapboxSpeechApiOptions,
-    private val resourceLoader: ResourceLoader
+    private val resourceLoader: ResourceLoader,
 ) {
 
-    suspend fun load(typeAndAnnouncement: TypeAndAnnouncement): Expected<Throwable, ByteArray> {
+    private val currentRequests = mutableSetOf<Long>()
+    private val downloadedInstructions = mutableSetOf<TypeAndAnnouncement>()
+    private val downloadedInstructionsLock = Any()
+    private val defaultScope = InternalJobControlFactory.createDefaultScopeJobControl().scope
+
+    suspend fun load(
+        voiceInstruction: VoiceInstructions,
+        onlyCache: Boolean
+    ): Expected<Throwable, ByteArray> {
         return runCatching {
-            val url = instructionUrl(typeAndAnnouncement.announcement, typeAndAnnouncement.type)
-            val response = resourceLoader.load(url)
+            val typeAndAnnouncement = VoiceInstructionsParser.parse(voiceInstruction)
+                .getValueOrElse { throw it }
+            val request = createRequest(typeAndAnnouncement).apply {
+                if (onlyCache) {
+                    networkRestriction = NetworkRestriction.DISALLOW_ALL
+                }
+            }
+            val response = resourceLoader.load(request)
             return processResponse(response)
         }.getOrElse {
             createError(it)
         }
     }
 
-    private suspend fun ResourceLoader.load(url: String) = load(ResourceLoadRequest(url))
+    fun triggerDownload(voiceInstructions: List<VoiceInstructions>) {
+        defaultScope.launch {
+            voiceInstructions.forEach { voiceInstruction ->
+                val typeAndAnnouncement = VoiceInstructionsParser.parse(voiceInstruction).value
+                if (typeAndAnnouncement != null && !hasTypeAndAnnouncement(typeAndAnnouncement)) {
+                    predownload(typeAndAnnouncement)
+                }
+            }
+        }
+    }
+
+    fun cancel() {
+        defaultScope.cancel()
+        currentRequests.forEach { resourceLoader.cancel(it) }
+    }
+
+    private fun hasTypeAndAnnouncement(typeAndAnnouncement: TypeAndAnnouncement): Boolean {
+        synchronized(downloadedInstructionsLock) {
+            return typeAndAnnouncement in downloadedInstructions
+        }
+    }
```

Comment:
Continuation of: https://github.com/mapbox/mapbox-navigation-android/pull/6547#discussion_r1053089938

> the Core SDK they won't start multiple identical requests.

Will it download again the instruction which was already downloaded?

Reply:
No, it will return the one from cache.

Reply:
why do you need `downloadedInstructions` then? Core SDK already handles everything, isn't it?

Reply:
It does. It's just an optimization on our side to avoid forming requests and accessing TileStore when not needed. And that might happen quite often because of the way the predownloading is triggered. Example:
1) trigger download of [A, B, C, D]
2) ...
3) trigger download of [D, E, F, G]
4) ...
5) trigger download of [F, G, H, I]

Triggering is based on timer and with the default settings the instructions to predownload will overlap often enough. And the settings are such because we want to avoid starting the download too late.